impl/src/expand.rs

```diff
@@ -58,18 +58,25 @@ fn impl_struct(input: Struct) -> TokenStream {
                     self.#source.as_dyn_error().backtrace()
                 }
             };
-            let combinator = if type_is_option(backtrace_field.ty) {
+            if &source_field.member == backtrace {
```

Comment:
The changes on the next 20 lines here can be replaced with a simpler change that does the same thing and minimizes repetition:

```diff
     };
-    let combinator = if type_is_option(backtrace_field.ty) {
+    let combinator = if source == backtrace {
+        source_backtrace
+    } else if type_is_option(backtrace_field.ty) {
         quote! {
```